x/datapool/keeper/pool_distribute.go

```diff
@@ -0,0 +1,238 @@
+		err = k.sendDepositToCurator(ctx, pool)
+		if err != nil {
+			return err
+		}
+
+		// distribute of revenue to each seller
+		err = k.executeRevenueDistribute(ctx, pool)
+		if err != nil {
+			return err
+		}
+	}
+	// delete this pool from instantRevenueDistribute
+	instantRevenueDistribute.RemovePreviousIndex(lastIndex)
+	k.SetInstantRevenueDistribute(ctx, instantRevenueDistribute)
+
+	return nil
+}
+
+// sendDepositToCurator returns to the Curator if the pool status is ACTIVE but the deposit has not yet been returned.
+func (k Keeper) sendDepositToCurator(ctx sdk.Context, pool *types.Pool) error {
+	if pool.Status != types.ACTIVE || pool.IsPaidDeposit {
+		return nil
+	}
+
+	curatorAddr, err := sdk.AccAddressFromBech32(pool.Curator)
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	poolAddr, err := sdk.AccAddressFromBech32(pool.PoolAddress)
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	err = k.bankKeeper.SendCoins(ctx, poolAddr, curatorAddr, sdk.NewCoins(pool.Deposit))
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	pool.IsPaidDeposit = true
+	k.SetPool(ctx, pool)
+
+	return nil
+}
+
+func (k Keeper) getAvailablePoolCoinAmount(ctx sdk.Context, pool *types.Pool) (*sdk.Int, error) {
+	poolAddr, err := sdk.AccAddressFromBech32(pool.PoolAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	poolBalance := k.bankKeeper.GetBalance(ctx, poolAddr, assets.MicroMedDenom)
+	if pool.IsPaidDeposit {
+		return &poolBalance.Amount, nil
+	} else {
+		amount := poolBalance.Amount.Sub(pool.Deposit.Amount)
+		return &amount, nil
+	}
+}
+
+func (k Keeper) executeRevenueDistribute(ctx sdk.Context, pool *types.Pool) error {
+	availablePoolCoinAmount, err := k.getAvailablePoolCoinAmount(ctx, pool)
+	if err != nil {
+		return err
+	}
+
+	// calculate the revenue to be sent to each seller
+	eachDistributionAmount := k.getEachDistributionAmount(pool)
+
+	poolAddress, err := sdk.AccAddressFromBech32(pool.GetPoolAddress())
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+
+	salesHistories := k.ListSalesHistory(ctx, pool.PoolId, pool.Round)
+	for _, history := range salesHistories {
+		// if there is no coin available, proceed no further.
+		if availablePoolCoinAmount.Equal(sdk.NewInt(0)) {
+			break
+		}
+
+		paidCoinAmount := history.PaidCoin.Amount
+		dataCount := sdk.NewInt(int64(history.DataCount()))
+		distributedAmount := eachDistributionAmount.Mul(dataCount)
+
+		// if it has already been distributed, pass it.
+		if distributedAmount.Equal(paidCoinAmount) {
+			continue
+		}
+
+		paymentAmount := distributedAmount.Sub(paidCoinAmount)
+		// If the transferable amount is less than the payable amount, it is replaced with the transferable amount.
+		if availablePoolCoinAmount.LT(paymentAmount) {
+			paymentAmount = *availablePoolCoinAmount
+		}
```

Comment:
Theoretically, I think the `availablePoolCoinAmount` cannot be less than `paymentAmount` if the `availablePoolCoinAmount` is not zero. Do you have any idea of this case?

Reply:
You're right!
In theory, this check is unnecessary.
Previously developed code was affected.
I fixed it. (https://github.com/medibloc/panacea-core/commit/da195915ca26f7e51121dc3cc674a287c06dcbdb)

Reply:
As you mentioned in TODO, BTW, we may have to discuss about error handling of this revenue distribution. I'll think about it too.

Reply:
Yes. We need a discussion.
Since this is an operation done in the endBlocker of the block chain, if an error occurs, block creation cannot be performed.
(It is likely that blocks will not continue to be created.)
I will discuss a few things soon.